tests/src/python/test_provider_hana.py

```diff
@@ -0,0 +1,361 @@
+                     '"time_field" TIME,' \
+                     '"datetime_field" TIMESTAMP)'
+        insert_sql = 'INSERT INTO "qgis_test"."date_time_type" ("id", "date_field", "time_field", "datetime_field") ' \
+                     'VALUES (?, ?, ?, ?)'
+        insert_args = [[1, '2004-03-04', '13:41:52', '2004-03-04 13:41:52']]
+        self.prepareTestTable('date_time_type', create_sql, insert_sql, insert_args)
+
+        vl = self.createVectorLayer('table="qgis_test"."date_time_type" sql=', 'testdatetimes')
+        self.assertTrue(vl.isValid())
+
+        fields = vl.dataProvider().fields()
+        self.assertEqual(fields.at(fields.indexFromName('date_field')).type(), QVariant.Date)
+        self.assertEqual(fields.at(fields.indexFromName('time_field')).type(), QVariant.Time)
+        self.assertEqual(fields.at(fields.indexFromName('datetime_field')).type(), QVariant.DateTime)
+
+        f = next(vl.getFeatures(QgsFeatureRequest()))
+
+        date_idx = vl.fields().lookupField('date_field')
+        self.assertIsInstance(f.attributes()[date_idx], QDate)
+        self.assertEqual(f.attributes()[date_idx], QDate(2004, 3, 4))
+        time_idx = vl.fields().lookupField('time_field')
+        self.assertIsInstance(f.attributes()[time_idx], QTime)
+        self.assertEqual(f.attributes()[time_idx], QTime(13, 41, 52))
+        datetime_idx = vl.fields().lookupField('datetime_field')
+        self.assertIsInstance(f.attributes()[datetime_idx], QDateTime)
+        self.assertEqual(f.attributes()[datetime_idx], QDateTime(QDate(2004, 3, 4), QTime(13, 41, 52)))
+
+    def testBinaryType(self):
+        create_sql = 'CREATE TABLE "qgis_test"."binary_type" ( ' \
+                     '"id" INTEGER NOT NULL PRIMARY KEY,' \
+                     '"blob" VARBINARY(114))'
+        insert_sql = 'INSERT INTO "qgis_test"."binary_type" ("id", "blob") VALUES (?, ?)'
+        insert_args = [[1, base64.b64encode(b'binvalue')], [2, None]]
+        self.prepareTestTable('binary_type', create_sql, insert_sql, insert_args)
+
+        vl = self.createVectorLayer('table="qgis_test"."binary_type" sql=', 'testbinary')
+        self.assertTrue(vl.isValid())
+
+        fields = vl.dataProvider().fields()
+        self.assertEqual(fields.at(fields.indexFromName('blob')).type(), QVariant.ByteArray)
+        self.assertEqual(fields.at(fields.indexFromName('blob')).length(), 114)
+
+        values = {feat['id']: feat['blob'] for feat in vl.getFeatures()}
+        expected = {1: QByteArray(b'YmludmFsdWU='), 2: QByteArray()}
+        self.assertEqual(values, expected)
+
+    def testBinaryTypeEdit(self):
+        create_sql = 'CREATE TABLE "qgis_test"."binary_type_edit" ( ' \
+                     '"id" INTEGER NOT NULL PRIMARY KEY,' \
+                     '"blob" VARBINARY(1000))'
+        insert_sql = 'INSERT INTO "qgis_test"."binary_type_edit" ("id", "blob") VALUES (?, ?)'
+        insert_args = [[1, base64.b64encode(b'bbb')]]
```

Comment:
I'd also expect here that we do a base64 decode.

Reply:
Done in 960bc2a.